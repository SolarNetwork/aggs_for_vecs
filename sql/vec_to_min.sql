-- smallint min lots
SELECT vec_to_min(smallints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- smallint min none
SELECT vec_to_min(smallints) FROM measurements WHERE sensor_id = -1;
-- smallint min one null
SELECT vec_to_min(smallints) FROM measurements WHERE sensor_id = 1;
-- smallint min array of nulls
SELECT vec_to_min(smallints) FROM measurements WHERE sensor_id = 2;
-- smallint min one not-null
SELECT vec_to_min(smallints) FROM measurements WHERE sensor_id = 4;
-- smallint min array of nulls and one other
SELECT vec_to_min(smallints) FROM measurements WHERE sensor_id IN (2, 4);

-- int min lots
SELECT vec_to_min(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- int min none
SELECT vec_to_min(ints) FROM measurements WHERE sensor_id = -1;
-- int min one null
SELECT vec_to_min(ints) FROM measurements WHERE sensor_id = 1;
-- int min array of nulls
SELECT vec_to_min(ints) FROM measurements WHERE sensor_id = 2;
-- int min one not-null
SELECT vec_to_min(ints) FROM measurements WHERE sensor_id = 4;
-- int min array of nulls and one other
SELECT vec_to_min(ints) FROM measurements WHERE sensor_id IN (2, 4);

-- bigint min lots
SELECT vec_to_min(bigints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- bigint min none
SELECT vec_to_min(bigints) FROM measurements WHERE sensor_id = -1;
-- bigint min one null
SELECT vec_to_min(bigints) FROM measurements WHERE sensor_id = 1;
-- bigint min array of nulls
SELECT vec_to_min(bigints) FROM measurements WHERE sensor_id = 2;
-- bigint min one not-null
SELECT vec_to_min(bigints) FROM measurements WHERE sensor_id = 4;
-- bigint min array of nulls and one other
SELECT vec_to_min(bigints) FROM measurements WHERE sensor_id IN (2, 4);

-- real min lots
SELECT vec_to_min(reals) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- real min none
SELECT vec_to_min(reals) FROM measurements WHERE sensor_id = -1;
-- real min one null
SELECT vec_to_min(reals) FROM measurements WHERE sensor_id = 1;
-- real min array of nulls
SELECT vec_to_min(reals) FROM measurements WHERE sensor_id = 2;
-- real min one not-null
SELECT vec_to_min(reals) FROM measurements WHERE sensor_id = 4;
-- real min array of nulls and one other
SELECT vec_to_min(reals) FROM measurements WHERE sensor_id IN (2, 4);

-- float8 min lots
SELECT vec_to_min(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- float8 min none
SELECT vec_to_min(floats) FROM measurements WHERE sensor_id = -1;
-- float8 min one null
SELECT vec_to_min(floats) FROM measurements WHERE sensor_id = 1;
-- float8 min array of nulls
SELECT vec_to_min(floats) FROM measurements WHERE sensor_id = 2;
-- float8 min one not-null
SELECT vec_to_min(floats) FROM measurements WHERE sensor_id = 4;
-- float8 min array of nulls and one other
SELECT vec_to_min(floats) FROM measurements WHERE sensor_id IN (2, 4);

-- numeric min lots
SELECT vec_to_min(nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- numeric min none
SELECT vec_to_min(nums) FROM measurements WHERE sensor_id = -1;
-- numeric min one null
SELECT vec_to_min(nums) FROM measurements WHERE sensor_id = 1;
-- numeric min array of nulls
SELECT vec_to_min(nums) FROM measurements WHERE sensor_id = 2;
-- numeric min one not-null
SELECT vec_to_min(nums) FROM measurements WHERE sensor_id = 4;
-- numeric min array of nulls and one other
SELECT vec_to_min(nums) FROM measurements WHERE sensor_id IN (2, 4);
-- numeric min data 01
-- results expected to match:
-- WITH di AS (
-- 	SELECT idx_i, min(val_i) AS min_i
-- 	FROM measurements2 d, unnest(d.data_i) WITH ORDINALITY AS a(val_i, idx_i)
-- 	GROUP BY idx_i
-- )
-- SELECT array_agg(min_i ORDER BY idx_i) AS data_i_min FROM di
SELECT vec_to_min(data_i) FROM measurements2;
