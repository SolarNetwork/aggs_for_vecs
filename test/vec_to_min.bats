load test_helper

@test "int16 min" {
  result="$(query "SELECT vec_to_min(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{1,2,2}" ]
}

# @test "int32 min" {
  # result="$(query "SELECT vec_to_min('{1,1,5,2,0}'::integer[])")";
  # [ "$result" = "0" ]
# }

# @test "int64 min" {
  # result="$(query "SELECT vec_to_min('{1,1,5,2,0}'::bigint[])")";
  # [ "$result" = "0" ]
# }

# @test "float4 min" {
  # result="$(query "SELECT vec_to_min('{1,1,5,2,0}'::real[])")";
  # [ "$result" = "0" ]
# }

@test "float8 min lots" {
  result="$(query "SELECT vec_to_min(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{1,2,2}" ]
}

@test "float8 min none" {
  result="$(query "SELECT vec_to_min(floats) FROM measurements WHERE sensor_id = -1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "float8 min one null" {
  result="$(query "SELECT vec_to_min(floats) FROM measurements WHERE sensor_id = 1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "float8 min array of nulls" {
  result="$(query "SELECT vec_to_min(floats) FROM measurements WHERE sensor_id = 2")";
  echo $result;
  [ "$result" = "{NULL,NULL,NULL}" ]
}

@test "float8 min one not-null" {
  result="$(query "SELECT vec_to_min(floats) FROM measurements WHERE sensor_id = 4")";
  echo $result;
  [ "$result" = "{1,NULL,2}" ]
}

@test "float8 min array of nulls and one other" {
  result="$(query "SELECT vec_to_min(floats) FROM measurements WHERE sensor_id IN (2, 4)")";
  echo $result;
  [ "$result" = "{1,NULL,2}" ]
}

@test "string min" {
  run query "SELECT vec_to_min(vals) FROM (VALUES (ARRAY['a']), (ARRAY['b'])) t(vals)"
  echo ${lines}
  [ "${lines[0]}" = "ERROR:  vec_to_min input must be array of SMALLINT, INTEGER, BIGINT, REAL, DOUBLE PRECISION, or NUMERIC" ]
}

@test "numeric min lots" {
  result="$(query "SELECT vec_to_min(nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{1.23,2.34,2.34}" ]
}

@test "numeric min none" {
  result="$(query "SELECT vec_to_min(nums) FROM measurements WHERE sensor_id = -1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "numeric min one null" {
  result="$(query "SELECT vec_to_min(nums) FROM measurements WHERE sensor_id = 1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "numeric min array of nulls" {
  result="$(query "SELECT vec_to_min(nums) FROM measurements WHERE sensor_id = 2")";
  echo $result;
  [ "$result" = "{NULL,NULL,NULL}" ]
}

@test "numeric min one not-null" {
  result="$(query "SELECT vec_to_min(nums) FROM measurements WHERE sensor_id = 4")";
  echo $result;
  [ "$result" = "{1.23,NULL,2.34}" ]
}

@test "numeric min array of nulls and one other" {
  result="$(query "SELECT vec_to_min(nums) FROM measurements WHERE sensor_id IN (2, 4)")";
  echo $result;
  [ "$result" = "{1.23,NULL,2.34}" ]
}