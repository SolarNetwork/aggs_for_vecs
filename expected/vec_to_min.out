-- smallint min lots
SELECT vec_to_min(smallints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_min 
------------
 {1,2,2}
(1 row)

-- smallint min none
SELECT vec_to_min(smallints) FROM measurements WHERE sensor_id = -1;
 vec_to_min 
------------
 
(1 row)

-- smallint min one null
SELECT vec_to_min(smallints) FROM measurements WHERE sensor_id = 1;
 vec_to_min 
------------
 
(1 row)

-- smallint min array of nulls
SELECT vec_to_min(smallints) FROM measurements WHERE sensor_id = 2;
    vec_to_min    
------------------
 {NULL,NULL,NULL}
(1 row)

-- smallint min one not-null
SELECT vec_to_min(smallints) FROM measurements WHERE sensor_id = 4;
 vec_to_min 
------------
 {1,NULL,2}
(1 row)

-- smallint min array of nulls and one other
SELECT vec_to_min(smallints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_min 
------------
 {1,NULL,2}
(1 row)

-- int min lots
SELECT vec_to_min(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_min 
------------
 {1,2,2}
(1 row)

-- int min none
SELECT vec_to_min(ints) FROM measurements WHERE sensor_id = -1;
 vec_to_min 
------------
 
(1 row)

-- int min one null
SELECT vec_to_min(ints) FROM measurements WHERE sensor_id = 1;
 vec_to_min 
------------
 
(1 row)

-- int min array of nulls
SELECT vec_to_min(ints) FROM measurements WHERE sensor_id = 2;
    vec_to_min    
------------------
 {NULL,NULL,NULL}
(1 row)

-- int min one not-null
SELECT vec_to_min(ints) FROM measurements WHERE sensor_id = 4;
 vec_to_min 
------------
 {1,NULL,2}
(1 row)

-- int min array of nulls and one other
SELECT vec_to_min(ints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_min 
------------
 {1,NULL,2}
(1 row)

-- bigint min lots
SELECT vec_to_min(bigints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_min 
------------
 {1,2,2}
(1 row)

-- bigint min none
SELECT vec_to_min(bigints) FROM measurements WHERE sensor_id = -1;
 vec_to_min 
------------
 
(1 row)

-- bigint min one null
SELECT vec_to_min(bigints) FROM measurements WHERE sensor_id = 1;
 vec_to_min 
------------
 
(1 row)

-- bigint min array of nulls
SELECT vec_to_min(bigints) FROM measurements WHERE sensor_id = 2;
    vec_to_min    
------------------
 {NULL,NULL,NULL}
(1 row)

-- bigint min one not-null
SELECT vec_to_min(bigints) FROM measurements WHERE sensor_id = 4;
 vec_to_min 
------------
 {1,NULL,2}
(1 row)

-- bigint min array of nulls and one other
SELECT vec_to_min(bigints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_min 
------------
 {1,NULL,2}
(1 row)

-- real min lots
SELECT vec_to_min(reals) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_min 
------------
 {1,2,2}
(1 row)

-- real min none
SELECT vec_to_min(reals) FROM measurements WHERE sensor_id = -1;
 vec_to_min 
------------
 
(1 row)

-- real min one null
SELECT vec_to_min(reals) FROM measurements WHERE sensor_id = 1;
 vec_to_min 
------------
 
(1 row)

-- real min array of nulls
SELECT vec_to_min(reals) FROM measurements WHERE sensor_id = 2;
    vec_to_min    
------------------
 {NULL,NULL,NULL}
(1 row)

-- real min one not-null
SELECT vec_to_min(reals) FROM measurements WHERE sensor_id = 4;
 vec_to_min 
------------
 {1,NULL,2}
(1 row)

-- real min array of nulls and one other
SELECT vec_to_min(reals) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_min 
------------
 {1,NULL,2}
(1 row)

-- float8 min lots
SELECT vec_to_min(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_min 
------------
 {1,2,2}
(1 row)

-- float8 min none
SELECT vec_to_min(floats) FROM measurements WHERE sensor_id = -1;
 vec_to_min 
------------
 
(1 row)

-- float8 min one null
SELECT vec_to_min(floats) FROM measurements WHERE sensor_id = 1;
 vec_to_min 
------------
 
(1 row)

-- float8 min array of nulls
SELECT vec_to_min(floats) FROM measurements WHERE sensor_id = 2;
    vec_to_min    
------------------
 {NULL,NULL,NULL}
(1 row)

-- float8 min one not-null
SELECT vec_to_min(floats) FROM measurements WHERE sensor_id = 4;
 vec_to_min 
------------
 {1,NULL,2}
(1 row)

-- float8 min array of nulls and one other
SELECT vec_to_min(floats) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_min 
------------
 {1,NULL,2}
(1 row)

-- numeric min lots
SELECT vec_to_min(nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
    vec_to_min    
------------------
 {1.23,2.34,2.34}
(1 row)

-- numeric min none
SELECT vec_to_min(nums) FROM measurements WHERE sensor_id = -1;
 vec_to_min 
------------
 
(1 row)

-- numeric min one null
SELECT vec_to_min(nums) FROM measurements WHERE sensor_id = 1;
 vec_to_min 
------------
 
(1 row)

-- numeric min array of nulls
SELECT vec_to_min(nums) FROM measurements WHERE sensor_id = 2;
    vec_to_min    
------------------
 {NULL,NULL,NULL}
(1 row)

-- numeric min one not-null
SELECT vec_to_min(nums) FROM measurements WHERE sensor_id = 4;
    vec_to_min    
------------------
 {1.23,NULL,2.34}
(1 row)

-- numeric min array of nulls and one other
SELECT vec_to_min(nums) FROM measurements WHERE sensor_id IN (2, 4);
    vec_to_min    
------------------
 {1.23,NULL,2.34}
(1 row)

-- numeric min data 01
-- results expected to match:
-- WITH di AS (
-- 	SELECT idx_i, min(val_i) AS min_i
-- 	FROM measurements2 d, unnest(d.data_i) WITH ORDINALITY AS a(val_i, idx_i)
-- 	GROUP BY idx_i
-- )
-- SELECT array_agg(min_i ORDER BY idx_i) AS data_i_min FROM di
SELECT vec_to_min(data_i) FROM measurements2;
                               vec_to_min                               
------------------------------------------------------------------------
 {0,-75585.95336,225.00002,49.500015,2531,0.09494743,-332.28107,2531,0}
(1 row)

